File: algorithms/sorting/bucket-sort/scala/BucketSort.scala
package algorithms.sorting.bucket

import scala.collection.mutable.ListBuffer

/**
 * Bucket Sort implementation.
 * Divides the input into several buckets, each of which is then sorted individually.
 */
object BucketSort {
  def sort(arr: Array[Int]): Array[Int] = {
    if (arr.length <= 1) {
      return arr.clone()
    }

    val n = arr.length
    val minVal = arr.min
    val maxVal = arr.max

    if (minVal == maxVal) {
      return arr.clone()
    }

    // Initialize buckets
    val buckets = Array.fill(n)(ListBuffer.empty[Int])
    val range = maxVal.toLong - minVal

    // Distribute elements into buckets
    for (x <- arr) {
      val index = ((x.toLong - minVal) * (n - 1) / range).toInt
      buckets(index) += x
    }

    // Sort each bucket and merge
    val result = new Array[Int](n)
    var k = 0
    for (bucket <- buckets) {
      if (bucket.nonEmpty) {
        val sortedBucket = bucket.sorted
        for (x <- sortedBucket) {
          result(k) = x
          k += 1
        }
      }
    }

    result
  }
}

File: algorithms/sorting/bucket-sort/kotlin/BucketSort.kt
package algorithms.sorting.bucket

/**
 * Bucket Sort implementation.
 * Divides the input into several buckets, each of which is then sorted individually.
 */
object BucketSort {
    fun sort(arr: IntArray): IntArray {
        if (arr.size <= 1) {
            return arr.copyOf()
        }

        val n = arr.size
        var min = arr[0]
        var max = arr[0]

        for (i in 1 until n) {
            if (arr[i] < min) min = arr[i]
            if (arr[i] > max) max = arr[i]
        }

        if (min == max) {
            return arr.copyOf()
        }

        // Initialize buckets
        val buckets = Array(n) { mutableListOf<Int>() }
        val range = max.toLong() - min

        // Distribute elements into buckets
        for (x in arr) {
            val index = ((x.toLong() - min) * (n - 1) / range).toInt()
            buckets[index].add(x)
        }

        // Sort each bucket and merge
        val result = IntArray(n)
        var k = 0
        for (bucket in buckets) {
            if (bucket.isNotEmpty()) {
                bucket.sort()
                for (x in bucket) {
                    result[k++] = x
                }
            }
        }

        return result
    }
}
